helpers/resource.c

```diff
@@ -122,3 +122,176 @@ size_t lai_read_resource(lai_nsnode_t *device, acpi_resource_t *dest) {
+        return info;
+    }
+}
+
+lai_api_error_t lai_resource_iterate(struct lai_resource_iterator *iterator){
+    LAI_ENSURE(iterator);
+    LAI_ENSURE(iterator->entry);
+
+    // Found our descriptor
+    struct lai_resource_header_info info = lai_get_header_info(iterator->entry);
+
+    // For some reason the resource size doesn't include the header, just like PkgLength
+    iterator->entry += info.skip_size;
+
+    uint8_t* entry = iterator->entry;
+    info = lai_get_header_info(entry);
+
+    if(info.variant == 0){
+        switch (info.type)
+        {
+        case ACPI_SMALL_FIXED_IO:
+            iterator->type = LAI_RESOURCE_IO;
+            iterator->base = (entry[1] | ((entry[2] & 0x3) << 8));
+            iterator->len = entry[3];
+            return LAI_ERROR_NONE;
+
+        case ACPI_SMALL_IRQ:
+            iterator->off = 0;
+            if(info.size == 2){
+                // No IRQ flags, use defaults
+                iterator->flags = ACPI_IRQ_ACTIVE_HIGH | ACPI_IRQ_EDGE | ACPI_IRQ_EXCLUSIVE;
+            } else if(info.size == 3){
+                iterator->flags = entry[3];
+            } else {
+                lai_panic("Unknown Small IRQ resource size: %02X", info.size);
+            }
+            return lai_resource_get_irq(iterator);
+
+        case ACPI_SMALL_END:
+            return LAI_ERROR_END_REACHED;
+
+        default:
+            lai_debug("undefined small resource, type is %02X, ignoring...",
+                            info.type);
+            break;
+        }
+
+        return LAI_ERROR_UNEXPECTED_RESULT;
+    } else if(info.variant == 1){
+        switch (info.type)
+        {
+        case ACPI_LARGE_IRQ:
+            iterator->off = 0;
+            iterator->flags = entry[3];
+            return lai_resource_get_irq(iterator);
+            break;
+        
+        default:
+            lai_debug("undefined large resource, type is %02X, ignoring...",
+                            info.type);
+            break;
+        }
+    } else {
+        lai_panic("Unknown Resource header variant %02X", info.variant);        
+    }
+    
+}
+
+enum lai_resource_type lai_resource_get_type(struct lai_resource_iterator *iterator){
+    LAI_ENSURE(iterator);
+    LAI_ENSURE(iterator->entry);
+    uint8_t* entry = iterator->entry;
+
+    struct lai_resource_header_info info = lai_get_header_info(entry);
+    if(info.variant == 0){
+        switch (info.type)
+        {
+        case ACPI_SMALL_IRQ:
+            return LAI_RESOURCE_IRQ;
+        case ACPI_SMALL_DMA:
+            return LAI_RESOURCE_DMA;
+        case ACPI_SMALL_IO:
+            return LAI_RESOURCE_IO;
+        case ACPI_SMALL_FIXED_IO:
+            return LAI_RESOURCE_IO;
+        case ACPI_SMALL_FIXED_DMA:
+            return LAI_RESOURCE_DMA;
+        case ACPI_SMALL_VENDOR:
+            return LAI_RESOURCE_VENDOR;
+        case ACPI_SMALL_END:
+            return LAI_NULL_NAME;
```

Comment:
`LAI_NULL_NAME` is from a completely unrelated `enum`.

Reply:
Resolved in f94abfb

Reply:
My bad